filter.go

```diff
@@ -43,31 +43,66 @@ const (
 	MapResultSkipDir
 )
 
-func Walk(ctx context.Context, p string, opt *WalkOpt, fn filepath.WalkFunc) error {
-	root, err := filepath.EvalSymlinks(p)
-	if err != nil {
-		return errors.WithStack(&os.PathError{Op: "resolve", Path: root, Err: err})
-	}
-	rootFI, err := os.Stat(root)
-	if err != nil {
-		return errors.WithStack(err)
+type filterFS struct {
+	fs  FS
+	opt *FilterOpt
+}
+
+func NewFilterFS(fs FS, opt *FilterOpt) FS {
+	if opt == nil {
+		return fs
 	}
-	if !rootFI.IsDir() {
-		return errors.WithStack(&os.PathError{Op: "walk", Path: root, Err: syscall.ENOTDIR})
+
+	if fs, ok := fs.(*filterFS); ok {
+		includes := append([]string{}, fs.opt.IncludePatterns...)
+		includes = append(includes, opt.IncludePatterns...)
+		excludes := append([]string{}, fs.opt.ExcludePatterns...)
+		excludes = append(excludes, opt.ExcludePatterns...)
+		follows := append([]string{}, fs.opt.FollowPaths...)
+		follows = append(follows, opt.FollowPaths...)
+		var mapFn MapFunc
+		if fs.opt.Map != nil && opt.Map != nil {
+			mapFn = func(path string, stat *types.Stat) MapResult {
+				result := fs.opt.Map(path, stat)
+				if result != MapResultKeep {
+					return result
+				}
+				return opt.Map(path, stat)
+			}
+		} else if fs.opt.Map != nil {
+			mapFn = fs.opt.Map
+		} else if opt.Map != nil {
+			mapFn = opt.Map
+		}
+		return &filterFS{fs, &FilterOpt{
+			IncludePatterns: includes,
+			ExcludePatterns: excludes,
+			FollowPaths:     follows,
+			Map:             mapFn,
+		}}
 	}
 
+	return &filterFS{fs, opt}
+}
+
+func (fs *filterFS) Open(p string) (io.ReadCloser, error) {
+	return fs.fs.Open(p)
+}
+
+func (fs *filterFS) Walk(ctx context.Context, target string, fn gofs.WalkDirFunc) error {
```

Comment:
This `Walk()` method here seems out of place. The signature for walk should be `Walk(fs, .., cb)`. FS should implement something that make this possible, eg. `ReadDir(name)` and `f.ReadDir()`

Reply:
The current semantics are difficult to express if we change to that model.

Imagine the following `ExcludePatterns`: (`/foo`, `!/foo/x`).

Currently, if we `Walk` this filesystem, the callback will only be called for `/foo` if `/foo/x` exists, otherwise it gets skipped.

If we try and express `ReadDir` like this, then `ReadDir("/")` will need to check for the existence of `/foo/x` to know if `/foo` should be returned. This would be quite a major rewrite of the include/exclude patterns logic, not sure if you think this would be worth it?

Reply:
Maybe the logic then needs to be that filter walks whole baseFS internally either on constructor/function-call or on the first use.

> The current semantics are difficult to express if we change to that model.

I see, but I don't see the benefit then of defining a `FilterFS` if it doesn't actually satisfy the `FS` interface.

Reply:
https://github.com/tonistiigi/fsutil/blob/36ef4d8c0dbba81b41fcfe7a03ba1f6ba56904f2/fs.go#L17-L20

It does though? `fsutil.FS` only defines `Open` and `Walk` for now.

I think switching to be more in match with `io/fs` is a good idea, but I'd propose that's out of scope for this PR - unless we want to make the argument to do a large PR to completely rework the internals of this package.